processor/metricextractprocessor/processor.go

```diff
@@ -0,0 +1,196 @@
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+// extractProcessor is a processor that extracts metrics from logs.
+type extractProcessor struct {
+	config   *Config
+	match    *expr.Expression
+	value    *expr.Expression
+	attrs    *expr.ExpressionMap
+	consumer consumer.Logs
+	logger   *zap.Logger
+}
+
+// newProcessor returns a new processor.
+func newProcessor(config *Config, consumer consumer.Logs, match, value *expr.Expression, attrs *expr.ExpressionMap, logger *zap.Logger) *extractProcessor {
+	return &extractProcessor{
+		config:   config,
+		match:    match,
+		value:    value,
+		attrs:    attrs,
+		consumer: consumer,
+		logger:   logger,
+	}
+}
+
+// Start starts the processor.
+func (e *extractProcessor) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+// Capabilities returns the consumer's capabilities.
+func (e *extractProcessor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+// Shutdown stops the processor.
+func (e *extractProcessor) Shutdown(_ context.Context) error {
+	return nil
+}
+
+// ConsumeLogs processes the logs.
+func (e *extractProcessor) ConsumeLogs(ctx context.Context, pl plog.Logs) error {
+	metrics := e.extractMetrics(pl)
+	if metrics.ResourceMetrics().Len() != 0 {
+		e.sendMetrics(ctx, metrics)
+	}
+
+	return e.consumer.ConsumeLogs(ctx, pl)
+}
+
+// extractMetrics extracts metrics from logs.
+func (e *extractProcessor) extractMetrics(pl plog.Logs) pmetric.Metrics {
+	recordGroups := expr.ConvertToResourceGroups(pl)
+
+	metrics := pmetric.NewMetrics()
+	for _, group := range recordGroups {
+		dataPoints := e.extractDataPoints(group.Records)
+		if dataPoints.Len() == 0 {
+			continue
+		}
+
+		metricResource := pmetric.NewResourceMetrics()
+		err := metricResource.Resource().Attributes().FromRaw(group.Resource)
+		if err != nil {
+			e.logger.Error("Failed to convert resource attributes", zap.Error(err))
+			continue
+		}
+
+		scopeMetrics := metricResource.ScopeMetrics().AppendEmpty()
+		scopeMetrics.Scope().SetName(typeStr)
+
+		metric := scopeMetrics.Metrics().AppendEmpty()
+		metric.SetName(e.config.MetricName)
+		metric.SetUnit(e.config.MetricUnit)
+
+		switch e.config.MetricType {
+		case gaugeDoubleType, gaugeIntType:
+			dataPoints.CopyTo(metric.SetEmptyGauge().DataPoints())
+		case counterDoubleType, counterIntType:
+			dataPoints.CopyTo(metric.SetEmptySum().DataPoints())
+		}
+
+		metricResource.CopyTo(metrics.ResourceMetrics().AppendEmpty())
+	}
+
+	return metrics
+}
+
+// extractDataPoints extracts data points from the records.
+func (e *extractProcessor) extractDataPoints(records []expr.Record) pmetric.NumberDataPointSlice {
+	dataPoints := pmetric.NewNumberDataPointSlice()
+
+	for _, record := range records {
+		if e.match.MatchRecord(record) {
```

Comment:
non blocking preference: would be a little nicer if we continued on the inverse of this to have a little less nesting

Reply:
That's a great idea. I'll gladly make this change.